File: mobile/lib/features/memory/domain/memory_read_model.dart
import 'package:memory_map/features/memory/domain/memory.dart';
import 'package:memory_map/features/memory/domain/memory_photo_preview.dart';

final class MemoryReadModel {
  const MemoryReadModel({
    required this.memory,
    this.previewPhoto,
  });

  factory MemoryReadModel.fromMemory(Memory memory) {
    return MemoryReadModel(memory: memory);
  }

  final Memory memory;
  final MemoryPhotoPreview? previewPhoto;

  bool get hasPreviewPhoto => previewPhoto != null;

  MemoryReadModel withMemoryMutation(Memory updatedMemory) {
    if (updatedMemory.id != memory.id) {
      throw ArgumentError('updatedMemory id must match memory id');
    }

    return MemoryReadModel(
      memory: updatedMemory,
      previewPhoto: previewPhoto,
    );
  }

  @override
  bool operator ==(Object other) {
    return identical(this, other) ||
        other is MemoryReadModel &&
            memory == other.memory &&
            previewPhoto == other.previewPhoto;
  }

  @override
  int get hashCode => Object.hash(memory, previewPhoto);

  @override
  String toString() {
    return 'MemoryReadModel(hasPreviewPhoto: $hasPreviewPhoto)';
  }
}

File: backend/src/main/java/memory_map/backend/memory/application/MemoryReadModel.java
package memory_map.backend.memory.application;

import memory_map.backend.memory.domain.Memory;

import java.util.Objects;

public record MemoryReadModel(

        Memory memory,

        MemoryPreviewPhoto previewPhoto

) {
    public MemoryReadModel {
        Objects.requireNonNull(memory, "memory must not be null");
    }

    public static MemoryReadModel withoutPreview(Memory memory) {
        return new MemoryReadModel(memory, null);
    }

    @Override
    public String toString() {
        return "MemoryReadModel[hasPreviewPhoto=%s]"
                .formatted(previewPhoto != null);
    }
}
